jib-core/src/main/java/com/google/cloud/tools/jib/tar/TarStreamBuilder.java

```diff
@@ -56,45 +48,35 @@ private void writeEntriesAsTarArchive(OutputStream tarByteStream) throws IOExcep
         new TarArchiveOutputStream(tarByteStream, StandardCharsets.UTF_8.name())) {
       // Enables PAX extended headers to support long file names.
       tarArchiveOutputStream.setLongFileMode(TarArchiveOutputStream.LONGFILE_POSIX);
-      for (Map.Entry<TarArchiveEntry, TarArchiveOutputStreamConsumer> entry :
-          archiveMap.entrySet()) {
+      for (Map.Entry<TarArchiveEntry, ByteSource> entry : archiveMap.entrySet()) {
         tarArchiveOutputStream.putArchiveEntry(entry.getKey());
-        entry.getValue().accept(tarArchiveOutputStream);
+        tarArchiveOutputStream.write(entry.getValue().read());
         tarArchiveOutputStream.closeArchiveEntry();
       }
     }
   }
 
   /**
-   * Adds an entry to the archive.
+   * Adds a {@link TarArchiveEntry} to the archive.
    *
-   * @param entry the entry to add.
+   * @param entry the {@link TarArchiveEntry}
    */
   public void addEntry(TarArchiveEntry entry) {
     archiveMap.put(
         entry,
-        tarArchiveOutputStream -> {
-          // Note that this will skip files that don't exist.
-          if (entry.isFile()) {
-            try (InputStream contentStream =
-                new BufferedInputStream(Files.newInputStream(entry.getFile().toPath()))) {
-              ByteStreams.copy(contentStream, tarArchiveOutputStream);
-            }
-          }
-        });
+        entry.isFile() ? MoreFiles.asByteSource(entry.getFile().toPath()) : ByteSource.empty());
```

Comment:
Using `Blob`, it would be `entry.isFile() ? Blobs.from(entry.getFile().toPath()) : Blobs.from(ignored -> {})`